device/src/keyboard.rs

```diff
@@ -0,0 +1,400 @@
+extern crate alloc;
+use core::fmt::Display;
+
+use crate::graphics::Graphics;
+use crate::graphics::{FONT_LARGE, FONT_MED, PADDING_TOP};
+use alloc::string::{String, ToString};
+use alloc::vec::Vec;
+use cst816s::TouchGesture;
+use cst816s::CST816S;
+use embedded_graphics::{
+    pixelcolor::Rgb565,
+    prelude::*,
+    primitives::*,
+    text::{Alignment, Text, TextStyleBuilder},
+};
+use embedded_graphics_framebuf::FrameBuf;
+use embedded_hal as hal;
+use esp_hal::timer::{self, timg::Timer};
+use esp_hal::Blocking;
+use frostsnap_core::schnorr_fun::frost::SecretShare;
+use fugit::Instant;
+use mipidsi::error::Error;
+use u8g2_fonts::U8g2TextStyle;
+
+const KEY_HEIGHT: u32 = 50;
+const BACKUP_LEFT_PADDING: u32 = 5;
+
+#[derive(Debug, Clone)]
+pub struct KeyboardKey {
+    label: KeyboardKeyType,
+    rectangle: Rectangle,
+}
+
+#[derive(Debug, Clone)]
+pub enum KeyboardKeyType {
+    Character(char),
+    String(String),
+}
+
+impl KeyboardKey {
+    pub fn new(point: Point, size: Size, label: KeyboardKeyType) -> Self {
+        let rectangle = Rectangle::new(point, size);
+        Self { label, rectangle }
+    }
+
+    pub fn rectangle(&self) -> Rectangle {
+        self.rectangle
+    }
+
+    pub fn label(&self) -> &KeyboardKeyType {
+        &self.label
+    }
+}
+
+impl Display for KeyboardKeyType {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            KeyboardKeyType::Character(c) => write!(f, "{}", c),
+            KeyboardKeyType::String(s) => write!(f, "{}", s),
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct Keyboard {
+    buffer: Vec<char>,
+    hrp: Option<String>,
+    last_touch: Option<Instant<u64, 1, 1_000_000>>,
+    touched_key: Option<KeyboardKey>,
+    keyboard_keys: Vec<Vec<KeyboardKey>>,
```

Comment:
I don't think we want this in the state of the keyboard. It seems like it could be a const. Furthermore I think you can figure out the rectangle and character just from the key's index or even just from the character itself in a functional way.